_DATA_TYPES/_Destructuring_assignment/main/main.js

```diff
@@ -0,0 +1,492 @@
+
+alert( user.name ); // Ilya
+
+/*
+Looping with .entries()
+
+We can use Object.entries(obj) method to loop over keys-and-values of an object
+*/
+
+let user = {
+    name: "John",
+    age: 30
+};
+
+// loop over keys and values
+
+for (let key of Object.entries(user)) {
+    alert( `${key}: ${value}`); // name: John, age: 30
+}
+
+
+// And the same for a map
+
+let user = new Map();
+user.set("name", "John");
+user.set("age", "30");
+
+for (let [key, value] of user.entries()) {
+    alert( `${key}: ${value}`); // name: John, age: 30
+}
+
+
+/*
+
+The rest '...'
+
+If we want not just to get first values, but also to gather all that follows - we can add one more parameter that gets
+"the rest" using three dots "..."
+*/
+
+
+let [name1, name2, ...rest] = ["Julius", "Caesar", "Consul", "of the Roman Republic"];
+
+alert(name1); // Julius
+alert(name2); // Caesar
+
+
+// Note that type of rest is Array
+
+alert(rest[0]); //Consul
+alert(rest[1]); // of the Roman Republic
+
+alert(rest.length); // 2
+
+/*
+The value of rest is the array of the remaining array elements. We can use any other variable name in place
+of rest, just make sure it has three dots before it and goes last in the destructuring assignment
+*/
+
+/*
+
+Default values
+
+If there are fewer values in the array than variables in the assignment, there will be NO error
+
+Absent values are considered undefined
+*/
+
+let [firstName, surname] = [];
+
+alert(firstName); //undefined
+alert(surname); // undefined
+
+
+// If we want a "default" value to replace the missing one, we can provide it using =
+
+// default values
+
+let [name = "Guest", surname = "Anonymous"] =  ["Julius"];
+
+alert(name); // Julius (from array)
+alert( surname ); //Anonymous (default used)
+
+
+/*
+Default values can be more complex expressions or even function calls. They are evaluates
+only if the value is not provided
+
+Here, we use prompt function for two defaults. But it will run only for the missing one
+*/
+
+// runs only prompt for surname
+
+let [name = prompt('name?'), surname = prompt('surname')] = ["Julius"];
+
+alert(name); //Julius
+alert(surname); //whatever prompt gets
+
+
+/*
+
+Object destructuring
+The basic syntax is
+
+*/
+
+ let {var1, var2} = {var1:..., var2...};
+
+ /*
+ We have an existing object at the right side, that we want to split into variables.
+
+ The left side contains a pattern for corresponding properties. In the simple case, that's a list of variable
+ names in {...}
+
+ e.g
+ */
+
+let options = {
+    title: "Menu",
+    width: 100,
+    height: 200
+};
+
+let { title, width, height } = options;
+
+alert(title); //Menu
+alert(width); // 100
+alert(height); // 200
+
+/*
+Properties options.title, options.width and options.height are assigned to the corresponding variables.
+The order does not matter
+
+This works too
+*/
+
+// changed the order of properties in let
+
+let {height, width, title} = { title: "Menu", height: 200, width: 100 }
```

Comment:
Semi-colons at the end of statements though not compulsory are generally a good practice.